service/worker/scanner/tasklist/handler.go

```diff
@@ -47,60 +48,45 @@ const scannerTaskListPrefix = "cadence-sys-tl-scanner"
 // with the assumption that the executor will schedule this task later
 //
 // Each loop of the handler proceeds as follows
-//    - Retrieve the next batch of tasks sorted by task_id for this task-list from persistence
+//    - Attempt to delete tasks up to the persisted ACK level.
 //    - If there are 0 tasks for this task-list, try deleting the task-list if its idle
-//    - If any of the tasks in the batch isn't expired, we are done. Since tasks are retrieved
-//      in sorted order, if one of the tasks isn't expired, chances are, none of the tasks above
-//      it are expired as well - so, we give up and wait for the next run
-//    - Delete the entire batch of tasks
-//    - If the number of tasks retrieved is less than batchSize, there are no more tasks in the task-list
+//    - If the number of tasks completed is less than the batchSize, there are no more tasks in the task-list
 //      Try deleting the task-list if its idle
-func (s *Scavenger) deleteHandler(key *taskListKey, state *taskListState) handlerStatus {
+func (s *Scavenger) deleteHandler(info *p.TaskListInfo) handlerStatus {
 	var err error
 	var nProcessed, nDeleted int
 
-	defer func() { s.deleteHandlerLog(key, state, nProcessed, nDeleted, err) }()
+	taskBatchSize := s.taskBatchSizeFn()
+	max := s.maxTasksPerJobFn()
 
-	for nProcessed < maxTasksPerJob {
-		resp, err1 := s.getTasks(key, taskBatchSize)
-		if err1 != nil {
-			err = err1
-			return handlerStatusErr
-		}
-
-		nTasks := len(resp.Tasks)
-		if nTasks == 0 {
-			s.tryDeleteTaskList(key, state)
-			return handlerStatusDone
-		}
+	defer func() { s.deleteHandlerLog(info, nProcessed, nDeleted, err) }()
 
-		for _, task := range resp.Tasks {
-			nProcessed++
-			if !s.isTaskExpired(task) {
-				return handlerStatusDone
-			}
+	for nProcessed < max {
+		nTasks, err := s.completeTasks(info, taskBatchSize)
+		if err == p.ErrPersistenceLimitExceeded {
+			s.logger.Info("scavenger.deleteHandler query was ratelimited; will retry")
+			return handlerStatusDefer
 		}
-
-		taskID := resp.Tasks[nTasks-1].TaskID
-		if _, err = s.completeTasks(key, taskID, nTasks); err != nil {
+		if err != nil {
+			s.logger.Error(fmt.Sprintf("Scavenger error completing tasks: %v", err))
 			return handlerStatusErr
 		}
-
+		nProcessed += nTasks
 		nDeleted += nTasks
 		if nTasks < taskBatchSize {
```

Comment:
After double thinking, I think this is wrong. Feel free to correct me:

`nTasks < taskBatchSize` is not a safe signal to delete a tasklist. `nTasks` is only the number of tasks that are already acked(polled). It doesn't mean the tasklist is expired, or there are no anymore valid tasks for the tasklist. 

Imagine this extreme case -- when `nTasks = ZERO`, but there are still tasks in the tasklist.  And if you delete the tasklist, then the remaining tasks become orphan(maybe that's the reason that you see so many orphan tasks). The problem is that, the task will be lost...You will see wrong behavior in Cadence workflow. 

However, like you said, the existing logic is also problematic that if see an non-expired task and then stops. So IMO, the deal way is this:
1. use `completeTasks` to complete acked tasks as much as we can. 
2. if `nTasks < taskBatchSize`, then do what the existing code is doing
   2.1 `getTasks(key, taskBatchSize)` to get all the tasks
   2.2 check if expired, and return if one of them is not(because we won't be able to delete the tasklist if so)
   2.3  complete(delete) the tasks up to the last taskID
   2.4. if 2.3 delete less than taskBatchSize, then the tasklist is safe to delete

Reply:
This call isn't truly a delete it's a "try delete" and only succeeds if the tasklist LastUpdated value is more than 48 hours old. If it isn't - meaning a history service is still handling the task list - then nothing happens. (And this was preexisting behavior for tryDeleteTaskList.)

I don't think scavenger should do anything at all above the acklevel, whether the task is expired or not.

Reply:
>  if the tasklist LastUpdated value is more than 48 hours old

This is not safe, because a task associated to the tasklist could have timeout greater than 48 hours. Simply deleting the tasklist will make the task become orphan. The method is preexisting, but the current way is safely checking all the tasks are expired. 

> I don't think scavenger should do anything at all above the acklevel, whether the task is expired or not.

This is another misunderstanding. The major purpose of this scavenger is to clean up the task and tasklist after expiration. This is a major design for SQL because SQL DBs are missing TTL feature. 

There are some context about the original design here. Sorry for not having enough documentation in Cadence yet. I am happy to provide more information if you have question. 

The main use case for TTL and deleting tasklist is that there is a feature in Cadence called `sticky tasklist`. The tasklist is auto-matcially created per instance per deployment(randomUUID). So we have to clean otherwise they will keep growing as people restart workers. Even though non-sticky tasklist could have the same issue, but less worrying than the sticky ones.

Reply:
What you say may be true, but is not related to a behavior introduced by this change. tryDeleteTaskList was always present and always had this behavior. What happens to tasks has changed; what happens to tasklists has not.

Reply:
Let me add some context behind this change in hopes it'll help reach mutual understanding.
* Our excess of orphans existed before I changed any code; this change is in response to that, among other things.
* We also had lingering tasks and tasklists for workflows that themselves were long expired and gone.

If I'm understanding your point, I think we're both agreed that completing tasks that are below the ack level is fine, regardless of expiration. The question remains what's correct to do above the ack level.

So the assertion you are making is it is safe to complete tasks that have expired even if the history service might be processing the task list? 

And secondly, that it is unsafe to delete a task list if it has unexpired tasks? If that's so, then I think at the very least the overall workflow state / timeout should also be taken as another factor. It should be safe to delete tasks and tasklists for a nonexistent workflow, regardless of the state of those items.

Reply:
I'm still trying to make sure I understand your point, so sorry for more questions...

In the Cassandra case, what happens when TTL is reached? Is there anything that preserves a tasklist entry due to unexpired tasks if nothing has happened to extend the TTL? How long is the TTL?

Reply:
> In the Cassandra case, what happens when TTL is reached ...

TTL is one day. It only applies to sticky tasklist. So normal tasklists will never get deleted...
I just realized what we did for SQL is quite different from Cassandra when reading the code --
Every time updating a sticky takslist(when updating ackLevel and rangeID), it will update the TTL again. 
Sticky tasklist is only for sticky decision tasks, they have very short timeout(usually seconds, default to 10s). So when TTL is reached, it means the sticky tasklist hasn't been used for last 24 hours, hence all tasks should have been expired. 

So from reading the code, we will never delete a normal tasklist in Cassandra. 

See more details in 
https://github.com/uber/cadence/blob/b78809c691b08c33bc2e11e6222b0e8fde75d7ee/common/persistence/cassandra/cassandraTaskPersistence.go#L291

Reply:
Just a little more background about tasklist table. It stores the metadata mainly for 
* rangeID: ownership management, and taskID management
  * everytime a matching host claim ownership of a tasklist, it must increase the value succesfully 
  * also used as base section of the taskID. E.g, if rangeID is 1, then allowed taskID ranged will be [100K, 2*100K-1], see more details in https://github.com/uber/cadence/blob/b78809c691b08c33bc2e11e6222b0e8fde75d7ee/service/matching/taskWriter.go#L129
* ackLevel: max taskID that can be safely deleted.

If there are non-expired tasks, but we delete the associated tasklist, the rangeID and ackLevel will be lost.  As a result, if   historySerive is calling matching service to add a new task, these two number will be starting from ZERO. This will cause some tasks will be overwritten as some taskIDs will be reused.

Reply:
Back to the current case of SQL tasklist scanner: It grabs a batch of tasks, and check if all of them expired, then delete the tasks and also the tasklist. 
This logic matches what we did in Cassandra for sticky tasklist. As sticky tasks only have short timeouts. 

However, this doesn't match what we did in Cassandra for normal tasklist. In Cassandra, most of the tasks will use TTL to clean up the tasks automatically. See https://github.com/uber/cadence/blob/b78809c691b08c33bc2e11e6222b0e8fde75d7ee/common/persistence/cassandra/cassandraTaskPersistence.go#L421 However, the  current SQL scanner will stop once it sees a task hasn't expired yet, without looking at other tasks. This could be a problem because in normal tasklist, decision task has a 1 year timeout by default. As a result, all the tasks will be only get deleted after one year. 

The existing code doesn't seem to have real issue of orphan tasks. I believe they are introduced by your change. By design tasklist will only get deleted only tasks are all deleted. 
However, I think your orphan tasks scanner is till nice to have. There could be bugs that we don't know. And should add some error logs into this scanner, because this is not expected to see.

Reply:
After reflecting on this overnight I was able to (I believe) understand your explanation. I'll bring back code checking for expiration. This will also bring back the "stuck scavenger" issue we were having for long-expiration tasks, so I'm also going to think about how to address that safely.